### PR TITLE
ASoC: rt700/711/715/1308: patches update

### PR DESCRIPTION
**Pull request by shumingfan**
ASoC: rt700/711/715: enable wake_capable
ASoC: rt700/711/1308-sdw: re-do io_init in cache_bypass mode when system resume

### CODE REVIEW

File: sound/soc/codecs/rt700.c [hunk 1]
```diff
@@ -1133,6 +1133,11 @@ int rt700_io_init(struct device *dev, struct sdw_slave *slave)
 	if (rt700->hw_init)
 		return 0;
 
+	if (rt700->first_init) {
+		regcache_cache_only(rt700->regmap, false);
+		regcache_cache_bypass(rt700->regmap, true);
```
Comment (plbossart) [inline sound/soc/codecs/rt700.c:1138]: sorry, I don't get why this is needed?

is this to force the regmap values to be synced?

if yes, why is this different from what we do in the resume case

````c
	regcache_cache_only(rt700->regmap, false);
	regcache_sync_region(rt700->regmap, 0x3000, 0x8fff);
	regcache_sync_region(rt700->regmap, 0x752010, 0x75206b);
````

I am not against, just having difficulties understanding the flow

File: sound/soc/codecs/rt700.c [hunk 2]
```diff
@@ -1198,10 +1201,12 @@ int rt700_io_init(struct device *dev, struct sdw_slave *slave)
...
 	regmap_write(rt700->regmap, RT700_SET_AUDIO_POWER_STATE, AC_PWRST_D3);
 
-	INIT_DELAYED_WORK(&rt700->jack_detect_work,
+	if (!rt700->first_init) {
+		INIT_DELAYED_WORK(&rt700->jack_detect_work,
 			rt700_jack_detect_handler);
-	INIT_DELAYED_WORK(&rt700->jack_btn_check_work,
+		INIT_DELAYED_WORK(&rt700->jack_btn_check_work,
 			rt700_btn_check_handler);
+	}
```
Comment (plbossart) [inline sound/soc/codecs/rt700.c:1209]: this sounds ok, but if we only do this on startup, then could we do this in the probe function instead?

it's a question, not an ask or recommendation.

File: sound/soc/codecs/rt700.c [hunk 3]
```diff
@@ -1210,6 +1215,11 @@ int rt700_io_init(struct device *dev, struct sdw_slave *slave)
 	if (rt700->hs_jack)
 		rt700_jack_init(rt700);
 
+	if (rt700->first_init)
+		regcache_cache_bypass(rt700->regmap, false);
```
Comment (plbossart) [inline sound/soc/codecs/rt700.c:1219]: same comment as above, I am not sure about the flow and interaction with resume and loss of sync cases.

File: sound/soc/codecs/rt711.c [hunk 4]
```diff
@@ -1174,6 +1174,11 @@ int rt711_io_init(struct device *dev, struct sdw_slave *slave)
 	if (rt711->hw_init)
 		return 0;
 
+	if (rt711->first_init) {
+		regcache_cache_only(rt711->regmap, false);
+		regcache_cache_bypass(rt711->regmap, true);
+	}
```
Comment (plbossart) [inline sound/soc/codecs/rt711.c:1180]: same question as for rt700

File: sound/soc/codecs/rt700.c (hunk 1 above)
Comment (plbossart) [inline sound/soc/codecs/rt700.c:1138]: and first_init is really 'first_boot'. This is something we done once, and the variable is not reset for suspend-resume, so the commit message is confusing to me.

**Comment by shumingfan** [comment]
@plbossart The main purpose is that the driver really re-do the io_init function when system resume.

When the codec suspend, the regmap needs to set cache_only mode.
Therefore, if the driver runs io_init at cache_only mode, it will not trigger HW to do EFUSE read or the calibration actually.
So, I use the first_init flag to check whether the cache_bypass enable or not and set cache_only to false.
It sets to cache_bypass to false when completed the io_init function.
The cache_bypass mode will not affect the cache data in regmap.
In the resume function, the regcache_sync could sync the registers back.
I think this method also could handle the case that the system cuts the power to codec in suspend mode.

The brief sequences:
- unattached
  - codec update_status
- enumeration 
  - enumeration_complete
- attached
  - codec update_status
    - io_init (enable cache_bypass that the driver doesn't affect the cache value in regmap)
  - initialization_complete	  
- codec resume 
  - regcache_sync

**Comment by plbossart** [comment]
Thanks @shumingfan for the explanations.
Now I get your points and indeed this would work for the first boot and suspend/resume.

There is one case however where your solution does not work. If the device loses sync or resets somehow while active, or even if we turn the power to the codec off while active (dangerous test!), the sequence will apply the initial settings directly in the hardware in io_init(), but there isn't a regmap_sync to re-apply the current settings.

That's where I feel we are missing something, the solution is not 100% complete IMHO.

**Comment by plbossart** [comment]
I will go ahead and merge though, that way we can test the suspend-resume cases.
We could track if we are resuming or not and add a regmap sync if we do the io_init outside of a suspend-resume cycle.

**Pull request merged**